projects/ADuCM3029_demo_cn0503/src/cn0503.c

```diff
@@ -3307,66 +3588,1420 @@ int32_t cn0503_prompt(struct cn0503_dev *dev)
+ *        Free any malloc'd memory.
+ *        Helper function for cn0503_impulse_response.
  * @param [in] dev - The device structure.
- * @param [out] code_rdy - false if new data is not ready;
- *                         true if data is ready.
- * @return SUCCESS in case of success, FAILURE otherwise.
  */
-static int32_t cn0503_update_code(struct cn0503_dev *dev, bool *code_rdy)
+void impulse_response_rollback(struct cn0503_dev *dev)
 {
-	int32_t ret;
-	uint16_t byte_count;
-	uint8_t no_data_samples = 0;
-	int8_t i;
-
-	*code_rdy = false;
-
-	ret = adpd410x_get_fifo_bytecount(dev->adpd4100_handler, &byte_count);
-	if(ret != SUCCESS)
-		return FAILURE;
-	for(i = 0; i < dev->active_slots; i++) {
-		if(dev->two_channel_slots & (1 << i))
-			no_data_samples += 2 * dev->data_sizes[i];
-		else
-			no_data_samples += dev->data_sizes[i];
+	int i;
+	struct cn0503_impulse_response *impresp = dev->impulse_response;
+	if (impresp == NULL)
+		return;
+	if (impresp->timer != NULL)
+		timer_remove(impresp->timer);
+	adpd410x_set_opmode(dev->adpd4100_handler, ADPD410X_STANDBY);
+	/** Rollback register values */
+	struct reg_config *registers = impresp->registers;
+
+	if (impresp->interrupt_set) {
+		gpio_setup_interrupt_callback(dev->adpd4100_handler->gpio1, NULL, NULL);
+	}
+	if (registers != NULL) {
+		for (i = impresp->nb_reg_writes-1; i >= 0; i--) {
+			// Write to rollback value register
+			adpd410x_reg_write(dev->adpd4100_handler, registers[i].addr,
+					   registers[i].rollback_value);
+		}
 	}
-	if(byte_count < no_data_samples)
-		return SUCCESS;
 
-	*code_rdy = true;
+	/** Rollback output data rate */
+	adpd410x_set_sampling_freq(dev->adpd4100_handler, CN0503_CODE_ODR_DEFAULT);
 
-	return adpd410x_get_data(dev->adpd4100_handler, dev->data_channel);
+	/** Rollback timeslot number */
+	adpd410x_set_last_timeslot(dev->adpd4100_handler, impresp->prev_active_slots);
+
+	/** Free malloc'd structures */
+	free(impresp->data);
+	free(impresp->averaged_data);
+	free(registers);
+	free(impresp);
+	dev->impulse_response = NULL;
 }
 
 /**
- * @brief Calculate ARAT equation if new data is available.
+ * @brief Set up and start a cycle of impulse response data collection.
+ *        Set integration offset, if necessary.
+ *        Clear FIFO, if necessary. Start data collection.
+ *        Helper function for cn0503_impulse_response.
  * @param [in] dev - The device structure.
- * @param [in] op_ptr - Pointer to the ASCII RPN equation.
- * @param [out] result - Result of the equation.
- * @return SUCCESS in case of success, FAILURE otherwise.
  */
-static int32_t cn0503_calc_ratio(struct cn0503_dev *dev, uint8_t **op_ptr,
-				 float *result)
+int32_t impulse_response_start_collection(struct cn0503_dev *dev)
 {
-	float a, b;
-	int8_t i = 0, op_id;
+	struct cn0503_impulse_response *impresp = dev->impulse_response;
+	uint16_t integ_offset_lower, integ_offset_upper, i = impresp->data_cycle;

```

Comment:
nitpick: is there a need here to use i instead of impresp->data_cycle directly?

Reply:
No, this was laziness. I can change i to impresp->data_cycle to avoid allocating more stack space.